protoc-gen-rust-grpc/src/grpc_rust_plugin.cc

```diff
@@ -0,0 +1,138 @@
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+ * IN THE SOFTWARE.
+ *
+ */
+
+#include "grpc_rust_generator.h"
+#include <google/protobuf/compiler/code_generator.h>
+#include <google/protobuf/compiler/plugin.h>
+#include <google/protobuf/compiler/rust/context.h>
+#include <google/protobuf/compiler/rust/crate_mapping.h>
+#include <google/protobuf/compiler/rust/naming.h>
+#include <vector>
+
+namespace protobuf = google::protobuf;
+namespace rust = google::protobuf::compiler::rust;
+
+static std::string ReconstructParameterList(
+    const std::vector<std::pair<std::string, std::string>> &options) {
+  std::string result;
+  for (const auto &[key, value] : options) {
+    if (!result.empty()) {
+      result += ",";
+    }
+    result += key + "=" + value;
+  }
+  return result;
+}
+
+class RustGrpcGenerator : public protobuf::compiler::CodeGenerator {
+public:
+  // Protobuf 5.27 released edition 2023.
+#if GOOGLE_PROTOBUF_VERSION >= 5027000
+  uint64_t GetSupportedFeatures() const override {
+    return Feature::FEATURE_PROTO3_OPTIONAL |
+           Feature::FEATURE_SUPPORTS_EDITIONS;
+  }
+  protobuf::Edition GetMinimumEdition() const override {
+    return protobuf::Edition::EDITION_PROTO2;
+  }
+  protobuf::Edition GetMaximumEdition() const override {
+    return protobuf::Edition::EDITION_2023;
+  }
+#else
+  uint64_t GetSupportedFeatures() const override {
+    return Feature::FEATURE_PROTO3_OPTIONAL;
+  }
+#endif
+
+  bool Generate(const protobuf::FileDescriptor *file,
+                const std::string &parameter,
+                protobuf::compiler::GeneratorContext *context,
+                std::string *error) const override {
+    // Return early to avoid creating an empty output file.
+    if (file->service_count() == 0) {
+      return true;
+    }
+    std::vector<std::pair<std::string, std::string>> options;
+    protobuf::compiler::ParseGeneratorParameter(parameter, &options);
+
+    // Filter out GRPC options.
+    std::vector<std::pair<std::string, std::string>> protobuf_options;
+    rust_grpc_generator::GrpcOpts grpc_opts;
+    for (auto opt : options) {
+      if (opt.first == "message_module_path") {
+        grpc_opts.message_module_path = opt.second;
+      } else {
+        protobuf_options.push_back(opt);
+      }
+    }
+
+    if (grpc_opts.message_module_path.empty()) {
+      grpc_opts.message_module_path = "self";
+    }
+
+    // The kernel isn't used by gRPC, it is there to pass Rust protobuf's
+    // validation.
+    protobuf_options.emplace_back("kernel", "upb");
```

Comment:
It seems awkward that you have to do this just to make our helper code happy. I think we should figure out a way to avoid needing to do this.

As far as I can tell, the only thing you need `google::protobuf::compiler::rust::Context` for is to be able to use the naming helper function `RsTypePath`. The reason that helper needs a `Context` is so that it can figure out how to refer to the symbol unambiguously based on the current position in the code (see logic [here](https://github.com/protocolbuffers/protobuf/blob/527b2ab41e785149a6189dba3052a210959b1372/src/google/protobuf/compiler/rust/naming.cc#L193)).

Reply:
Yes, I needed to create a context to call `protobuf::compiler::rust::RsTypePath`. The context has information about the crate mapping that it uses for calling `GetCrateName` [here](https://github.com/protocolbuffers/protobuf/blob/527b2ab41e785149a6189dba3052a210959b1372/src/google/protobuf/compiler/rust/naming.cc#L223C23-L223C36).

Reply:
I've copied the symbol resolution code and the crate mapping parsing code to avoid using `rust::Context`.